File: backend/app/routers/memory.py
from fastapi import APIRouter, HTTPException, Query, BackgroundTasks
from typing import Optional, List
from pydantic import BaseModel

from app.services.memory import (
    CoreMemory,
    MemoryService,
    MemoryExtractor,
    MemoryRetriever,
)
from app.services.memory.types import (
    CoreMemoryUpdate,
    RecallMemoryCreate,
    ArchivalMemoryCreate,
    MemorySearchResult,
    MemoryStats,
    ProcessConversationRequest,
    MemoryExtractionResult,
    MemoryConfig,
    MemoryCategory,
    MemoryContext,
)
from app.services.memory.service import memory_service
from app.services.memory.auto_capture import AutoCaptureService
from app.services.memory.auto_recall import AutoRecallService
from app.services.memory.auto_forget import AutoForgetService

router = APIRouter(prefix="/memory", tags=["memory"])

auto_capture_service = AutoCaptureService()
auto_recall_service = AutoRecallService()
auto_forget_service = AutoForgetService()


@router.get("/core", response_model=CoreMemory)
async def get_core_memory(user_id: str = Query(default="default")):
    """Get user's core memory (profile)."""
    memory = await memory_service.get_core_memory(user_id)
    if not memory:
        return CoreMemory(user_id=user_id)
    return memory


@router.put("/core", response_model=CoreMemory)
async def update_core_memory(
    update: CoreMemoryUpdate,
    user_id: str = Query(default="default"),
):
    """Update user's core memory (profile)."""
    return await memory_service.update_core_memory(user_id, update)


@router.get("/recall")
async def get_recall_memories(
    user_id: str = Query(default="default"),
    limit: int = Query(default=100, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
):
    """Get user's recall memories (conversation history)."""
    memories = await memory_service.get_recall_memories(user_id, limit, offset)
    return [
        {
            "memory_id": m.memory_id,
            "user_id": m.user_id,
            "session_id": m.session_id,
            "timestamp": m.timestamp.isoformat() if hasattr(m.timestamp, 'isoformat') else m.timestamp,
            "content": m.content,
            "importance_score": m.importance_score,
            "access_count": m.access_count,
        }
        for m in memories
    ]


@router.post("/recall", response_model=dict)
async def create_recall_memory(
    data: RecallMemoryCreate,
    user_id: str = Query(default="default"),
):
    """Create a new recall memory."""
    memory = await memory_service.create_recall_memory(user_id, data)
    return {
        "memory_id": memory.memory_id,
        "timestamp": memory.timestamp.isoformat(),
        "importance_score": memory.importance_score,
    }


@router.delete("/recall/{memory_id}")
async def delete_recall_memory(memory_id: str):
    """Delete a specific recall memory."""
    success = await memory_service.delete_recall_memory(memory_id)
    if not success:
        raise HTTPException(status_code=404, detail="Memory not found")
    return {"success": True}


@router.get("/archival")
async def get_archival_memories(
    user_id: str = Query(default="default"),
    limit: int = Query(default=100, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
):
    """Get user's archival memories (knowledge base)."""
    memories = await memory_service.get_archival_memories(user_id, limit, offset)
    return [
        {
            "memory_id": m.memory_id,
            "user_id": m.user_id,
            "content_type": m.content_type,
            "title": m.title,
            "content": m.content,
            "source_papers": m.source_papers,
            "tags": m.tags,
            "created_at": m.created_at.isoformat() if hasattr(m.created_at, 'isoformat') else m.created_at,
            "last_accessed": m.last_accessed.isoformat() if hasattr(m.last_accessed, 'isoformat') else m.last_accessed,
        }
        for m in memories
    ]


@router.post("/archival", response_model=dict)
async def create_archival_memory(
    data: ArchivalMemoryCreate,
    user_id: str = Query(default="default"),
):
    """Create a new archival memory (note/insight/summary)."""
    memory = await memory_service.create_archival_memory(user_id, data)
    return {
        "memory_id": memory.memory_id,
        "created_at": memory.created_at.isoformat(),
        "content_type": memory.content_type,
    }


@router.delete("/archival/{memory_id}")
async def delete_archival_memory(memory_id: str):
    """Delete a specific archival memory."""
    success = await memory_service.delete_archival_memory(memory_id)
    if not success:
        raise HTTPException(status_code=404, detail="Memory not found")
    return {"success": True}


@router.post("/search", response_model=List[MemorySearchResult])
async def search_memories(
    query: str,
    user_id: str = Query(default="default"),
    top_k: int = Query(default=5, ge=1, le=20),
):
    """Search memories by semantic similarity."""
    return await memory_service.search_memories(query, user_id, top_k)


@router.get("/stats", response_model=MemoryStats)
async def get_memory_stats(user_id: str = Query(default="default")):
    """Get memory statistics for a user."""
    return await memory_service.get_memory_stats(user_id)


@router.delete("/clear")
async def clear_all_memories(user_id: str = Query(default="default")):
    """Clear all memories for a user."""
    success = await memory_service.clear_all_memories(user_id)
    if not success:
        raise HTTPException(status_code=500, detail="Failed to clear memories")
    return {"success": True}


@router.delete("/clear/core")
async def clear_core_memory(user_id: str = Query(default="default")):
    """Clear core memory (profile) for a user."""
    success = await memory_service.clear_core_memory(user_id)
    if not success:
        raise HTTPException(status_code=500, detail="Failed to clear core memory")
    return {"success": True}


@router.delete("/clear/recall")
async def clear_recall_memories(user_id: str = Query(default="default")):
    """Clear all recall memories for a user."""
    success = await memory_service.clear_recall_memories(user_id)
    if not success:
        raise HTTPException(status_code=500, detail="Failed to clear recall memories")
    return {"success": True}


@router.delete("/clear/archival")
async def clear_archival_memories(user_id: str = Query(default="default")):
    """Clear all archival memories for a user."""
    success = await memory_service.clear_archival_memories(user_id)
    if not success:
        raise HTTPException(status_code=500, detail="Failed to clear archival memories")
    return {"success": True}


@router.get("/context")
async def get_memory_context(
    query: str,
    user_id: str = Query(default="default"),
):
    """Get relevant memory context for a query."""
    context = await memory_service.build_context_for_query(query, user_id)
    return {"context": context}


@router.get("/profile")
async def get_user_profile(user_id: str = Query(default="default")):
    """Get user profile summary."""
    summary = await memory_service.get_user_profile_summary(user_id)
    return {"profile": summary}


@router.get("/recommended-skills")
async def get_recommended_skills(user_id: str = Query(default="default")):
    """Get recommended skills based on user's frequently used skills."""
    skills = await memory_service.get_recommended_skills(user_id)
    return {"skills": skills}


@router.post("/process-conversation")
async def process_conversation(
    data: ProcessConversationRequest,
    background_tasks: BackgroundTasks,
    user_id: str = Query(default="default"),
):
    """Process a conversation and extract memories in the background.
    
    This endpoint immediately returns a task ID and processes the memory
    extraction asynchronously to avoid blocking the main thread.
    
    Args:
        extract: If True, extract profile data from conversation and update CoreMemory.
                 If False, only store as RecallMemory without profile extraction.
    """
    import uuid
    task_id = str(uuid.uuid4())
    
    background_tasks.add_task(
        memory_service.process_conversation,
        user_id=user_id,
        session_id=data.session_id,
        user_message=data.user_message,
        assistant_message=data.assistant_message,
        extract=data.extract,
    )
    
    return {
        "status": "processing",
        "task_id": task_id,
        "message": "Conversation queued for memory processing",
    }


@router.get("/config", response_model=MemoryConfig)
async def get_memory_config(user_id: str = Query(default="default")):
    """Get memory configuration for a user."""
    return await memory_service.get_memory_config(user_id)


@router.put("/config", response_model=MemoryConfig)
async def update_memory_config(
    config: MemoryConfig,
    user_id: str = Query(default="default"),
):
    """Update memory configuration for a user."""
    success = await memory_service.save_memory_config(user_id, config)
    if not success:
        raise HTTPException(status_code=500, detail="Failed to save config")
    return config


class StoreMemoryRequest(BaseModel):
    text: str
    category: Optional[MemoryCategory] = None
    importance: Optional[float] = None


@router.post("/store")
async def store_memory(
    request: StoreMemoryRequest,
    user_id: str = Query(default="default"),
):
    """Manually store a memory."""
    memory = await auto_capture_service.manual_store(
        user_id=user_id,
        text=request.text,
        category=request.category,
        importance=request.importance,
    )
    if not memory:
        raise HTTPException(status_code=500, detail="Failed to store memory")
    return memory


@router.get("/recall-search", response_model=List[MemorySearchResult])
async def recall_memories(
    query: str,
    limit: int = Query(default=5, ge=1, le=2000),
    user_id: str = Query(default="default"),
    min_score: float = Query(default=0.0, ge=0.0, le=1.0),
):
    """Search and recall memories by query."""
    config = MemoryConfig(
        auto_recall=True,
        recall_top_k=limit,
        recall_min_score=min_score,
    )
    context = await auto_recall_service.recall_for_query(query, user_id, config)
    return context.memories


@router.get("/context-result", response_model=MemoryContext)
async def get_memory_context_result(
    query: str,
    limit: int = Query(default=5, ge=1, le=2000),
    user_id: str = Query(default="default"),
    min_score: float = Query(default=0.0, ge=0.0, le=1.0),
):
    """Get memory context with both memories and context string."""
    config = MemoryConfig(
        auto_recall=True,
        recall_top_k=limit,
        recall_min_score=min_score,
    )
    return await auto_recall_service.recall_for_query(query, user_id, config)


@router.delete("/forget")
async def forget_memories(
    memory_id: Optional[str] = None,
    user_id: str = Query(default="default"),
):
    """Forget a specific memory or all auto-created memories."""
    if memory_id:
        success = await auto_forget_service.forget_memory(memory_id, user_id)
    else:
        deleted = await auto_forget_service.forget_all_auto_created(user_id)
        success = deleted > 0
    
    return {"success": success}


@router.delete("/forget/{memory_id}")
async def forget_memory_by_id(
    memory_id: str,
    user_id: str = Query(default="default"),
):
    """Forget a specific memory by ID."""
    success = await auto_forget_service.forget_memory(memory_id, user_id)
    if not success:
        raise HTTPException(status_code=404, detail="Memory not found")
    return {"success": True}


@router.post("/cleanup")
async def cleanup_expired_memories(user_id: str = Query(default="default")):
    """Clean up expired low-importance memories."""
    config = MemoryConfig()
    deleted = await auto_forget_service.cleanup_expired_memories(user_id, config)
    return {"deleted": deleted}

File: src/services/memory.ts
import type {
  CoreMemory,
  CoreMemoryUpdate,
  RecallMemory,
  ArchivalMemory,
  ArchivalMemoryCreate,
  MemoryStats,
  MemorySearchResult,
  ProcessConversationRequest,
  MemoryExtractionResult,
  MemoryConfig,
  MemoryCategory,
  MemoryContextResult,
} from '@/types/memory'
import { API_BASE_URL } from './config'

const API_BASE = `${API_BASE_URL}/api/memory`

async function fetchJson<T>(url: string, options?: RequestInit): Promise<T> {
  const response = await fetch(url, {
    headers: {
      'Content-Type': 'application/json',
      ...options?.headers,
    },
    ...options,
  })
  
  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`)
  }
  
  return response.json()
}

export const memoryService = {
  async getCoreMemory(): Promise<CoreMemory | null> {
    return fetchJson<CoreMemory>(`${API_BASE}/core`)
  },

  async updateCoreMemory(update: CoreMemoryUpdate): Promise<CoreMemory> {
    return fetchJson<CoreMemory>(`${API_BASE}/core`, {
      method: 'PUT',
      body: JSON.stringify(update),
    })
  },

  async getRecallMemories(limit: number = 100, offset: number = 0): Promise<RecallMemory[]> {
    return fetchJson<RecallMemory[]>(`${API_BASE}/recall?limit=${limit}&offset=${offset}`)
  },

  async deleteRecallMemory(memoryId: string): Promise<boolean> {
    const response = await fetchJson<{ success: boolean }>(`${API_BASE}/recall/${memoryId}`, {
      method: 'DELETE',
    })
    return response.success
  },

  async getArchivalMemories(limit: number = 100, offset: number = 0): Promise<ArchivalMemory[]> {
    return fetchJson<ArchivalMemory[]>(`${API_BASE}/archival?limit=${limit}&offset=${offset}`)
  },

  async createArchivalMemory(data: ArchivalMemoryCreate): Promise<ArchivalMemory> {
    const response = await fetchJson<{ memory_id: string; created_at: string; content_type: string }>(
      `${API_BASE}/archival`,
      {
        method: 'POST',
        body: JSON.stringify(data),
      }
    )
    return {
      memory_id: response.memory_id,
      user_id: '',
      content_type: data.content_type || 'note',
      title: data.title || '',
      content: data.content,
      source_papers: data.source_papers || [],
      tags: data.tags || [],
      created_at: response.created_at,
      last_accessed: response.created_at,
    }
  },

  async deleteArchivalMemory(memoryId: string): Promise<boolean> {
    const response = await fetchJson<{ success: boolean }>(`${API_BASE}/archival/${memoryId}`, {
      method: 'DELETE',
    })
    return response.success
  },

  async searchMemories(query: string, topK: number = 5): Promise<MemorySearchResult[]> {
    return fetchJson<MemorySearchResult[]>(`${API_BASE}/search?query=${encodeURIComponent(query)}&top_k=${topK}`, {
      method: 'POST',
    })
  },

  async getMemoryStats(): Promise<MemoryStats> {
    return fetchJson<MemoryStats>(`${API_BASE}/stats`)
  },

  async clearAllMemories(): Promise<boolean> {
    const response = await fetchJson<{ success: boolean }>(`${API_BASE}/clear`, {
      method: 'DELETE',
    })
    return response.success
  },

  async clearCoreMemory(): Promise<boolean> {
    const response = await fetchJson<{ success: boolean }>(`${API_BASE}/clear/core`, {
      method: 'DELETE',
    })
    return response.success
  },

  async clearRecallMemories(): Promise<boolean> {
    const response = await fetchJson<{ success: boolean }>(`${API_BASE}/clear/recall`, {
      method: 'DELETE',
    })
    return response.success
  },

  async clearArchivalMemories(): Promise<boolean> {
    const response = await fetchJson<{ success: boolean }>(`${API_BASE}/clear/archival`, {
      method: 'DELETE',
    })
    return response.success
  },

  async getMemoryContext(query: string): Promise<string> {
    const response = await fetchJson<{ context: string }>(`${API_BASE}/context?query=${encodeURIComponent(query)}`)
    return response.context
  },

  async getUserProfile(): Promise<string> {
    const response = await fetchJson<{ profile: string }>(`${API_BASE}/profile`)
    return response.profile
  },

  async getRecommendedSkills(): Promise<string[]> {
    const response = await fetchJson<{ skills: string[] }>(`${API_BASE}/recommended-skills`)
    return response.skills
  },

  async processConversation(data: ProcessConversationRequest): Promise<MemoryExtractionResult> {
    return fetchJson<MemoryExtractionResult>(`${API_BASE}/process-conversation`, {
      method: 'POST',
      body: JSON.stringify(data),
    })
  },

  async getMemoryConfig(): Promise<MemoryConfig> {
    return fetchJson<MemoryConfig>(`${API_BASE}/config`)
  },

  async updateMemoryConfig(config: Partial<MemoryConfig>): Promise<MemoryConfig> {
    return fetchJson<MemoryConfig>(`${API_BASE}/config`, {
      method: 'PUT',
      body: JSON.stringify(config),
    })
  },

  async storeMemory(
    text: string,
    category?: MemoryCategory,
    importance?: number
  ): Promise<RecallMemory> {
    return fetchJson<RecallMemory>(`${API_BASE}/store`, {
      method: 'POST',
      body: JSON.stringify({ text, category, importance }),
    })
  },

  async recallMemories(query: string, limit?: number): Promise<MemorySearchResult[]> {
    return fetchJson<MemorySearchResult[]>(
      `${API_BASE}/recall-search?query=${encodeURIComponent(query)}&limit=${limit || 5}`
    )
  },

  async forgetMemory(memoryId?: string): Promise<boolean> {
    const url = memoryId ? `${API_BASE}/forget/${memoryId}` : `${API_BASE}/forget`
    const response = await fetchJson<{ success: boolean }>(url, {
      method: 'DELETE',
    })
    return response.success
  },

  async getMemoryContextResult(query: string): Promise<MemoryContextResult> {
    return fetchJson<MemoryContextResult>(
      `${API_BASE}/context-result?query=${encodeURIComponent(query)}`
    )
  },

  async cleanupExpiredMemories(): Promise<{ deleted: number }> {
    return fetchJson<{ deleted: number }>(`${API_BASE}/cleanup`, {
      method: 'POST',
    })
  },
}
